clients/privacy-center/components/RequestModal.tsx

```diff
@@ -0,0 +1,293 @@
+  Input,
+  FormErrorMessage,
+} from '@fidesui/react';
+
+import { useFormik } from 'formik';
+
+import type { AlertState } from '../types/AlertState';
+
+import config from '../config/config.json';
+
+export const useRequestModal = () => {
+  const [isOpen, setIsOpen] = useState(false);
+  const [openAction, setOpenAction] = useState<string | null>(null);
+
+  const onOpen = (action: string) => {
+    setOpenAction(action);
+    setIsOpen(true);
+  };
+
+  const onClose = () => {
+    setIsOpen(false);
+    setOpenAction(null);
+  };
+
+  return { isOpen, onClose, onOpen, openAction };
+};
+
+const useRequestForm = ({
+  onClose,
+  action,
+  setAlert,
+}: {
+  onClose: () => void;
+  action: typeof config.actions[0] | null;
+  setAlert: (state: AlertState) => void;
+}) => {
+  const [isLoading, setIsLoading] = useState(false);
+  const formik = useFormik({
+    initialValues: {
+      name: '',
+      email: '',
+      phone: '',
+    },
+    onSubmit: async (values) => {
+      if (!action) {
+        // somehow we've reached a broken state, return
+        return;
+      }
+
+      setIsLoading(true);
+      const host =
+        process.env.NODE_ENV === 'development'
+          ? config.fidesops_host_development
+          : config.fidesops_host_production;
+      const body = [
+        {
+          identity: {
+            email: values.email,
+            phone_number: values.phone,
+            // enable this when name field is supported on the server
+            // name: values.name
+          },
+          policy_key: action.policy_key,
+        },
+      ];
+
+      const handleError = () => {
+        setAlert({
+          status: 'error',
+          description: 'Your request has failed. Please try again.',
+        });
+        onClose();
+      };
+
+      try {
+        const response = await fetch(`${host}/privacy-request`, {
+          method: 'POST',
+          headers: {
+            Accept: 'application/json',
+            'Content-Type': 'application/json',
+          },
+          body: JSON.stringify(body),
+        });
+
+        if (!response.ok) {
+          handleError();
+          return;
+        }
+
+        const data = await response.json();
+
+        if (data.succeeded.length) {
+          setAlert({
+            status: 'success',
+            description:
+              'Your request was successful, please await further instructions.',
+          });
+        } else {
+          handleError();
+        }
+      } catch (error) {
+        handleError();
+        return;
+      }
+
+      onClose();
+    },
+    validate: (values) => {
+      if (!action) return {};
+      const errors: {
+        name?: string;
+        email?: string;
+        phone?: string;
+      } = {};
+
+      if (!values.email && action.identity_inputs.email === 'required') {
+        errors.email = 'Required';
+      } else if (
+        values.email &&
+        !/^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}$/i.test(values.email)
```

Comment:
I'd like for us to get better at documenting regex's in general. This is obviously validating an email address but can we add a comment here breaking down exactly what'll match and what won't? Then if we found this anywhere, a link to the original source.

It looks as though the regex here might not cover all the cases we need too, for example we're not including the tilde `~` character in the set of valid chars.

![Screenshot 2022-03-10 at 14 40 31](https://user-images.githubusercontent.com/1667340/157741807-177c018d-2c4a-4506-b9b5-3cda843ee38f.png)

I understand this particular topic is hard to pin down, Wikipedia's resource looks well informed: https://en.wikipedia.org/wiki/Email_address#Local-part_normalization

I suspect it might be better here to err on the side of letting through emails that _might_ be invalid, since the worst case scenario is that it gets run through Fidesops as a privacy request and if the email is invalid then nothing will be found.

Reply:
OWASP's input validation cheatsheet has a pretty good section on email address validation which might be helpful here, @seanpreston?

Link: https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html#email-address-validation

Basically, it's surprisingly very difficult to do it right and the extended validation should be offloaded to a mail server if possible, but some basic initial validation could look like this:

* The email address contains two parts, separated with an @ symbol.
* The email address does not contain dangerous characters (such as backticks, single or double quotes, or null bytes).
   * Exactly which characters are dangerous will depend on how the address is going to be used (echoed in page, inserted into database, etc).
* The domain part contains only letters, numbers, hyphens (-) and periods (.).
* The email address is a reasonable length:
   * The local part (before the @) should be no more than 63 characters.
   * The total length should be no more than 254 characters.

Reply:
I've been baited into this conversation by @seanpreston asking for strong opinions

My strong opinion is that we _shouldn't_ try to do this kind of validation ourselves; `<input type="email">` exists and the browsers handle it...

Reply:
Hmmmm. Front-end validation is trivial to bypass anyway.

Reply:
For context, the regex in this PR comes from the Formik docs (Formik is an extremely popular React form management library).

Given @neville's opinion, I'll remove email validation for now, and we can add it back in a future PR if needed.